// boltz-core@5.0.0/dist/lib/musig/Musig.js
import { secp256k1 } from '@noble/curves/secp256k1.js';
import { hex } from '@scure/base';
import { Session, keyAggExport, keyAggregate, nonceAggregate, nonceGen, } from '@scure/btc-signer/musig2.js';
import { equalBytes } from '@scure/btc-signer/utils.js';
const assertParticipantIndex = (index, length) => {
    if (!Number.isInteger(index) || index < 0 || index >= length) {
        throw new Error('index out of range');
    }
    return index;
};
const assertUniquePublicKeys = (publicKeys) => {
    const seen = new Set();
    for (const key of publicKeys) {
        const encoded = hex.encode(key);
        if (seen.has(encoded)) {
            throw new Error(`duplicate public key ${encoded}`);
        }
        seen.add(encoded);
    }
};
const assertPublicKeyLength33 = (publicKey) => {
    if (publicKey.length !== 33) {
        throw new Error(`public key must be 33 bytes, got ${publicKey.length}`);
    }
};
const assertPublicKeys = (publicKeys) => {
    for (const publicKey of publicKeys) {
        assertPublicKeyLength33(publicKey);
    }
    assertUniquePublicKeys(publicKeys);
};
const normalizePublicKeys = (publicKeys) => {
    const keys = [...publicKeys];
    assertPublicKeys(keys);
    return Object.freeze(keys);
};
const allPartialsPresent = (partials) => partials.every((partial) => partial !== null);
const findKeyIndex = (publicKeys, targetKey) => publicKeys.findIndex((key) => equalBytes(targetKey, key));
const indexOfPublicKeyOrIndex = (publicKeys, publicKeyOrIndex) => {
    if (typeof publicKeyOrIndex === 'number') {
        return assertParticipantIndex(publicKeyOrIndex, publicKeys.length);
    }
    const index = findKeyIndex(publicKeys, publicKeyOrIndex);
    if (index === -1) {
        throw new Error(`could not find index of public key ${hex.encode(publicKeyOrIndex)}`);
    }
    return assertParticipantIndex(index, publicKeys.length);
};
class MusigBase {
    state;
    constructor(state) {
        this.state = state;
    }
    get aggPubkey() {
        return this.state.aggPubkey;
    }
    get internalKey() {
        return this.state.internalKey;
    }
    get publicKeys() {
        return this.state.publicKeys;
    }
    get myPublicKey() {
        return this.state.myPublicKey;
    }
    get myIndex() {
        return this.state.myIndex;
    }
    get tweak() {
        return this.state.tweak;
    }
    get numParticipants() {
        return this.state.publicKeys.length;
    }
    hasPublicKey = (publicKey) => {
        return findKeyIndex(this.state.publicKeys, publicKey) !== -1;
    };
    indexOfPublicKey = (publicKey) => {
        return indexOfPublicKeyOrIndex(this.state.publicKeys, publicKey);
    };
}
class MusigWithMessageBase extends MusigBase {
    get msg() {
        return this.state.msg;
    }
}
class MusigWithNonceBase extends MusigWithMessageBase {
    get publicNonce() {
        return this.state.nonce.public;
    }
}
class MusigWithSessionBase extends MusigWithNonceBase {
    verifyPartial = (publicKeyOrIndex, signature) => {
        const index = indexOfPublicKeyOrIndex(this.state.publicKeys, publicKeyOrIndex);
        return this.state.session.partialSigVerify(signature, Array.from(this.state.pubNonces), index);
    };
    addPartial = (publicKeyOrIndex, signature) => {
        if (!this.verifyPartial(publicKeyOrIndex, signature)) {
            throw new Error('invalid partial signature');
        }
        const index = indexOfPublicKeyOrIndex(this.state.publicKeys, publicKeyOrIndex);
        this.state.partialSignatures[index] = signature;
        return this;
    };
    addPartials = (partials) => {
        for (const [publicKeyOrIndex, signature] of partials) {
            this.addPartial(publicKeyOrIndex, signature);
        }
        return this;
    };
}
class MusigWithOurPartialSignatureBase extends MusigWithSessionBase {
    get ourPartialSignature() {
        return this.state.ourSignature;
    }
}
class MusigKeyAgg extends MusigBase {
    xonlyTweakAdd = (tweak) => {
        if (this.state.tweak) {
            throw new Error('musig key already tweaked');
        }
        return new MusigKeyAgg({
            ...this.state,
            aggPubkey: aggregateKeys(this.state.publicKeys, tweak),
            tweak,
        });
    };
    message = (msg) => {
        return new MusigWithMessage({ ...this.state, msg });
    };
}
class MusigWithMessage extends MusigWithMessageBase {
    generateNonce = () => {
        const nonce = nonceGen(this.state.myPublicKey, this.state.privateKey, this.state.aggPubkey, this.state.msg);
        return new MusigWithNonce({ ...this.state, nonce });
    };
}
class MusigWithNonce extends MusigWithNonceBase {
    aggregateNonces = (nonces) => {
        const noncesToUse = Array.from(nonces);
        const ourNonceEntry = noncesToUse.find(([keyCmp]) => equalBytes(this.state.myPublicKey, keyCmp));
        if (ourNonceEntry === undefined) {
            noncesToUse.push([this.state.myPublicKey, this.publicNonce]);
        }
        else if (!equalBytes(ourNonceEntry[1], this.publicNonce)) {
            throw new Error('nonce for our public key does not match our generated nonce');
        }
        if (this.state.publicKeys.length !== noncesToUse.length) {
            throw new Error('number of nonces != number of public keys');
        }
        const nonceByKey = new Map();
        for (const [key, nonce] of noncesToUse) {
            const encoded = hex.encode(key);
            if (nonceByKey.has(encoded)) {
                throw new Error(`duplicate nonce for public key ${encoded}`);
            }
            nonceByKey.set(encoded, nonce);
        }
        const ordered = [];
        for (const key of this.state.publicKeys) {
            const nonce = nonceByKey.get(hex.encode(key));
            if (nonce === undefined) {
                throw new Error(`could not find nonce for public key ${hex.encode(key)}`);
            }
            ordered.push(nonce);
        }
        return this.aggregateNoncesOrdered(ordered);
    };
    aggregateNoncesOrdered = (nonces) => {
        if (this.state.publicKeys.length !== nonces.length) {
            throw new Error('number of nonces != number of public keys');
        }
        if (!equalBytes(nonces[this.state.myIndex], this.state.nonce.public)) {
            throw new Error('our nonce is at incorrect index');
        }
        const orderedNonces = Object.freeze([...nonces]);
        return new MusigNoncesAggregated({
            ...this.state,
            pubNonces: orderedNonces,
            aggregatedNonce: nonceAggregate([...orderedNonces]),
        });
    };
}
class MusigNoncesAggregated extends MusigWithNonceBase {
    initializeSession = () => {
        const session = new Session(this.state.aggregatedNonce, Array.from(this.state.publicKeys), this.state.msg, this.state.tweak ? [this.state.tweak] : [], this.state.tweak ? [true] : []);
        return new MusigSession({
            ...this.state,
            session,
            partialSignatures: Array(this.state.publicKeys.length).fill(null),
        });
    };
}
class MusigSession extends MusigWithSessionBase {
    signPartial = () => {
        const sig = this.state.session.sign(this.state.nonce.secret, this.state.privateKey, true);
        const partialSignatures = [...this.state.partialSignatures];
        partialSignatures[this.state.myIndex] = sig;
        return new MusigSigned({
            ...this.state,
            partialSignatures,
            ourSignature: sig,
        });
    };
}
class MusigSigned extends MusigWithOurPartialSignatureBase {
    aggregatePartials = () => {
        if (!allPartialsPresent(this.state.partialSignatures)) {
            throw new Error('not all partial signatures are set');
        }
        return this.state.session.partialSigAgg(this.state.partialSignatures);
    };
}
const aggregateKeys = (publicKeys, tweak) => {
    assertPublicKeys(publicKeys);
    return keyAggExport(keyAggregate([...publicKeys], tweak ? [tweak] : [], tweak ? [true] : []));
};
const create = (privateKey, publicKeys) => {
    if (publicKeys.length < 2) {
        throw new Error('need at least 2 keys to aggregate');
    }
    const normalizedPublicKeys = normalizePublicKeys(publicKeys);
    const myPublicKey = secp256k1.getPublicKey(privateKey);
    const myIndex = findKeyIndex(normalizedPublicKeys, myPublicKey);
    if (myIndex === -1) {
        throw new Error('our key is not in publicKeys');
    }
    const aggPubkey = aggregateKeys(normalizedPublicKeys);
    return new MusigKeyAgg({
        privateKey,
        myPublicKey,
        publicKeys: normalizedPublicKeys,
        myIndex: assertParticipantIndex(myIndex, normalizedPublicKeys.length),
        aggPubkey,
        internalKey: aggPubkey,
    });
};
export { aggregateKeys, create, MusigKeyAgg, MusigWithMessage, MusigWithNonce, MusigNoncesAggregated, MusigSession, MusigSigned, };
