// boltz-core@5.0.0/dist/lib/swap/SwapTreeSerializer.js
import { hex } from '@scure/base';
import { assignTreeProbabilities, sortTree } from "./TreeSort.js";
const serializeLeaf = (leaf) => ({
    version: leaf.version,
    output: hex.encode(leaf.output),
});
const deserializeLeaf = (leaf) => ({
    version: leaf.version,
    output: hex.decode(leaf.output),
});
export const serializeSwapTree = (tree) => {
    const res = {
        claimLeaf: serializeLeaf(tree.claimLeaf),
        refundLeaf: serializeLeaf(tree.refundLeaf),
    };
    if (tree.covenantClaimLeaf !== undefined) {
        res.covenantClaimLeaf = serializeLeaf(tree.covenantClaimLeaf);
    }
    return res;
};
export const serializeFundingAddressTree = (tree) => ({
    refundLeaf: serializeLeaf(tree.refundLeaf),
});
export const deserializeSwapTree = (tree) => {
    const parsed = typeof tree === 'string' ? JSON.parse(tree) : tree;
    const res = {
        claimLeaf: deserializeLeaf(parsed.claimLeaf),
        refundLeaf: deserializeLeaf(parsed.refundLeaf),
        covenantClaimLeaf: parsed.covenantClaimLeaf !== undefined
            ? deserializeLeaf(parsed.covenantClaimLeaf)
            : undefined,
    };
    return {
        ...res,
        tree: sortTree(assignTreeProbabilities(res)),
    };
};
export const deserializeFundingAddressTree = (tree) => {
    const parsed = typeof tree === 'string' ? JSON.parse(tree) : tree;
    const refundLeaf = deserializeLeaf(parsed.refundLeaf);
    return {
        refundLeaf,
        tree: refundLeaf,
    };
};
