// boltz-core@5.0.0/dist/lib/liquid/swap/TaprootUtils.js
import { hex } from '@scure/base';
import { Transaction } from 'liquidjs-lib';
import { findScriptPath as liquidFindScriptPath, tapLeafHash as liquidTapLeafHash, } from 'liquidjs-lib/src/bip341.js';
import { taggedHash } from 'liquidjs-lib/src/crypto.js';
import { toXOnly } from "../../swap/TaprootUtils.js";
import { secp } from "../init.js";
const convertLeaf = (leaf) => ({
    version: leaf.version,
    scriptHex: hex.encode(leaf.output),
});
export const hashForWitnessV1 = (network, details, tx, index, leafHash, hashType = Transaction.SIGHASH_DEFAULT) => {
    return tx.hashForWitnessV1(index, details.map((out) => out.script), details.map((out) => ({
        value: out.value,
        asset: out.asset,
    })), hashType, network.genesisBlockHash, leafHash);
};
export const tapLeafHash = (leaf) => liquidTapLeafHash(convertLeaf(leaf));
export const tapBranchHash = (a, b) => taggedHash('TapBranch/elements', Buffer.concat([a, b]));
export const tapTweakHash = (publicKey, tweak) => taggedHash('TapTweak/elements', Buffer.concat([toXOnly(publicKey), tweak]));
export function toHashTree(scriptTree) {
    if (!Array.isArray(scriptTree)) {
        return { hash: tapLeafHash(scriptTree) };
    }
    const hashes = [toHashTree(scriptTree[0]), toHashTree(scriptTree[1])];
    hashes.sort((a, b) => a.hash.compare(b.hash));
    const [left, right] = hashes;
    return {
        hash: tapBranchHash(left.hash, right.hash),
        left,
        right,
    };
}
export const tweakMusig = (musig, tree) => {
    const tweak = toHashTree(tree).hash;
    return musig.xonlyTweakAdd(tapTweakHash(Buffer.from(musig.aggPubkey), tweak));
};
export const createControlBlock = (hashTree, leaf, internalKey) => {
    const leafHash = tapLeafHash(leaf);
    const path = liquidFindScriptPath(hashTree, leafHash);
    // In Taproot, the script path from a leaf to the root consists of sibling hashes.
    // For single-leaf trees (e.g. FundingAddressTree), there are no siblings, so the
    // path is empty and the leaf hash equals the root hash. This is valid because the
    // control block only needs the internal key - no additional path elements are required
    if (path === undefined ||
        (path.length === 0 && !hashTree.hash.equals(leafHash))) {
        throw new Error('leaf not in tree');
    }
    const outputKey = secp.ecc.xOnlyPointAddTweak(internalKey, tapTweakHash(internalKey, hashTree.hash));
    if (outputKey === null) {
        throw new Error('output key is null');
    }
    return Buffer.concat([Buffer.from([leaf.version | outputKey.parity]), internalKey].concat(path));
};
